File: mcp-code-intelligence-nodejs/src/context/curated-context-service.ts
/**
 * KSA-160: Curated Context Service — NL query → parallel search → RRF merge → budget allocation.
 */

import Database from 'better-sqlite3';
import { QueryAnalyzer } from './query-analyzer.js';
import { RRFMerger } from './rrf-merger.js';
import { BudgetAllocator, AllocatedResult } from './budget-allocator.js';
import { SymbolResolver } from '../graph/symbol-resolver.js';
import { GraphTraverser } from '../graph/traverser.js';
import { QueryLayer } from '../query/query-layer.js';
import {
  CuratedContextParams, CuratedContextResponse, ContextSection, ContextItem,
  SourceWeights, MergedResult
} from './types.js';

export class CuratedContextService {
  private analyzer: QueryAnalyzer;
  private merger: RRFMerger;
  private allocator: BudgetAllocator;
  private db: Database.Database;
  private queryLayer: QueryLayer;
  private traverser: GraphTraverser;
  private resolver: SymbolResolver;

  constructor(
    db: Database.Database,
    queryLayer: QueryLayer,
    traverser: GraphTraverser,
    resolver: SymbolResolver
  ) {
    this.analyzer = new QueryAnalyzer();
    this.merger = new RRFMerger();
    this.allocator = new BudgetAllocator();
    this.db = db;
    this.queryLayer = queryLayer;
    this.traverser = traverser;
    this.resolver = resolver;
  }

  /** Execute curated context search with NL query. */
  async getContext(params: CuratedContextParams): Promise<CuratedContextResponse> {
    const startTime = Date.now();
    const {
      query,
      max_tokens = 4000,
      include_source = true,
      include_memory = true,
      include_graph = true,
      source_weights
    } = params;

    // 1. Analyze query
    const analysis = this.analyzer.analyze(query);

    // 2. Parallel search
    const [codeResults, memoryResults] = await Promise.all([
      include_source ? this.searchCode(analysis) : Promise.resolve({ source: 'code', results: [] }),
      include_memory ? this.searchMemory(analysis) : Promise.resolve({ source: 'memory', results: [] })
    ]);

    // 3. Graph expansion (depends on code results)
    let graphResults = { source: 'graph', results: [] as any[] };
    if (include_graph && codeResults.results.length > 0) {
      graphResults = await this.expandGraph(codeResults.results.slice(0, 5));
    }

    // 4. Merge with RRF
    const merged = this.merger.merge(
      { code: codeResults, memory: memoryResults, graph: graphResults },
      source_weights
    );

    // 5. Allocate token budget
    const allocated = this.allocator.allocate(merged, max_tokens);

    // 6. Format response
    const sections = this.formatSections(allocated);
    const tokensUsed = allocated.reduce((sum, r) => sum + r.tokens, 0) + 100;

    return {
      query,
      sections,
      metadata: {
        tokens_used: tokensUsed,
        tokens_budget: max_tokens,
        sources_queried: [
          ...(include_source ? ['code'] : []),
          ...(include_memory ? ['memory'] : []),
          ...(include_graph ? ['graph'] : [])
        ],
        total_candidates: codeResults.results.length + memoryResults.results.length + graphResults.results.length,
        results_returned: allocated.length,
        execution_time_ms: Date.now() - startTime
      }
    };
  }

  private async searchCode(analysis: { ftsQuery: string; symbolCandidates: string[] }): Promise<{ source: string; results: any[] }> {
    try {
      const ftsResults = this.queryLayer.searchCode(analysis.ftsQuery, 30);

      // Also try direct symbol resolution for candidates
      const symbolResults: any[] = [];
      for (const candidate of analysis.symbolCandidates.slice(0, 5)) {
        const resolved = this.resolver.resolve(candidate);
        for (const sym of resolved.slice(0, 3)) {
          symbolResults.push({
            id: sym.id,
            name: sym.name,
            kind: sym.kind,
            file: sym.filePath,
            line: sym.line,
            signature: null
          });
        }
      }

      // Combine FTS + symbol results (mini-RRF)
      const combined = this.miniRRF(
        ftsResults.map(r => ({
          id: undefined,
          name: r.name,
          kind: r.kind,
          file: r.filePath,
          line: r.startLine,
          signature: r.signature,
          source_code: null
        })),
        symbolResults
      );

      return { source: 'code', results: combined.slice(0, 20) };
    } catch {
      return { source: 'code', results: [] };
    }
  }

  private async searchMemory(analysis: { ftsQuery: string; keywords: string[] }): Promise<{ source: string; results: any[] }> {
    try {
      // Search memory/KB using FTS on knowledge entries
      const query = analysis.keywords.slice(0, 5).join(' ');
      const rows = this.db.prepare(`
        SELECT id, content, summary, type, tags, created_at
        FROM knowledge_entries
        WHERE id IN (
          SELECT rowid FROM knowledge_fts WHERE knowledge_fts MATCH ?
        )
        ORDER BY created_at DESC
        LIMIT 10
      `).all(query) as any[];

      return {
        source: 'memory',
        results: rows.map(r => ({
          id: r.id,
          name: r.summary || r.content?.substring(0, 50) || 'entry',
          kind: r.type || 'memory',
          content: r.summary || r.content?.substring(0, 200),
          file: undefined,
          line: undefined
        }))
      };
    } catch {
      // KB tables may not exist
      return { source: 'memory', results: [] };
    }
  }

  private async expandGraph(topSymbols: any[]): Promise<{ source: string; results: any[] }> {
    const expanded: any[] = [];
    const seen = new Set<string>();

    for (const symbol of topSymbols) {
      try {
        const startNode = this.traverser.resolveNode(symbol.name);
        if (!startNode) continue;

        const results = this.traverser.traverse(startNode, {
          edgeTypes: ['calls', 'imports', 'inherits'],
          nodeTypes: [],
          direction: 'both',
          maxDepth: 1,
          maxResults: 5
        });

        for (const r of results) {
          const key = `${r.node.name}:${r.node.filePath}`;
          if (seen.has(key)) continue;
          seen.add(key);

          expanded.push({
            id: r.node.id,
            name: r.node.name,
            kind: r.node.kind,
            file: r.node.filePath,
            line: r.node.startLine,
            relationship: `${r.edgeType} ${symbol.name}`
          });
        }
      } catch {
        // Skip if symbol not in graph
      }
    }

    return { source: 'graph', results: expanded };
  }

  private miniRRF(listA: any[], listB: any[]): any[] {
    const k = 60;
    const scores = new Map<string, { score: number; item: any }>();

    for (let i = 0; i < listA.length; i++) {
      const key = listA[i].name + ':' + (listA[i].file || '');
      scores.set(key, { score: 1 / (k + i), item: listA[i] });
    }

    for (let i = 0; i < listB.length; i++) {
      const key = listB[i].name + ':' + (listB[i].file || '');
      if (scores.has(key)) {
        scores.get(key)!.score += 1 / (k + i);
      } else {
        scores.set(key, { score: 1 / (k + i), item: listB[i] });
      }
    }

    return Array.from(scores.values())
      .sort((a, b) => b.score - a.score)
      .map(e => e.item);
  }

  private formatSections(allocated: AllocatedResult[]): ContextSection[] {
    const bySource = new Map<string, ContextItem[]>();

    for (const item of allocated) {
      const source = (item.sources?.[0] || 'code') as 'code' | 'memory' | 'graph';
      if (!bySource.has(source)) bySource.set(source, []);

      bySource.get(source)!.push({
        name: item.name,
        kind: item.kind,
        file: item.file,
        line: item.line,
        relevance: item.relevance_score,
        detail: item.detail,
        content: item.content,
        relationship: item.relationship
      });
    }

    const sections: ContextSection[] = [];
    const titleMap: Record<string, string> = {
      code: 'Code Symbols',
      memory: 'Knowledge Base',
      graph: 'Related (Graph)'
    };

    for (const [source, items] of bySource) {
      sections.push({
        title: titleMap[source] || source,
        source: source as 'code' | 'memory' | 'graph',
        items
      });
    }

    return sections;
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/context/curated_context_service.py
"""Curated Context Service — NL query → parallel search → RRF merge → budget allocation. KSA-171."""

from __future__ import annotations

import sqlite3
import time
from typing import Any

from ..graph.symbol_resolver import SymbolResolver
from ..graph.traverser import GraphTraverser
from ..query import QueryLayer
from .budget_allocator import AllocatedResult, BudgetAllocator
from .query_analyzer import QueryAnalyzer
from .rrf_merger import RRFMerger
from .types import (
    ContextItem, ContextSection, CuratedContextParams, CuratedContextResponse,
)


class CuratedContextService:
    """NL query → parallel search → RRF merge → budget allocation."""

    def __init__(
        self,
        conn: sqlite3.Connection,
        query_layer: QueryLayer,
        traverser: GraphTraverser,
        resolver: SymbolResolver,
    ) -> None:
        self._analyzer = QueryAnalyzer()
        self._merger = RRFMerger()
        self._allocator = BudgetAllocator()
        self._conn = conn
        self._query_layer = query_layer
        self._traverser = traverser
        self._resolver = resolver

    def get_context(self, params: CuratedContextParams) -> CuratedContextResponse:
        """Execute curated context search with NL query."""
        start_time = time.time()

        # 1. Analyze query
        analysis = self._analyzer.analyze(params.query)

        # 2. Search sources
        code_results = self._search_code(analysis) if params.include_source else {"source": "code", "results": []}
        memory_results = self._search_memory(analysis) if params.include_memory else {"source": "memory", "results": []}

        # 3. Graph expansion (depends on code results)
        graph_results: dict[str, Any] = {"source": "graph", "results": []}
        if params.include_graph and code_results["results"]:
            graph_results = self._expand_graph(code_results["results"][:5])

        # 4. Merge with RRF
        merged = self._merger.merge(
            {"code": code_results, "memory": memory_results, "graph": graph_results},
            params.source_weights,
        )

        # 5. Allocate token budget
        allocated = self._allocator.allocate(merged, params.max_tokens)

        # 6. Format response
        sections = self._format_sections(allocated)
        tokens_used = sum(r.tokens for r in allocated) + 100

        elapsed_ms = int((time.time() - start_time) * 1000)
        sources_queried: list[str] = []
        if params.include_source:
            sources_queried.append("code")
        if params.include_memory:
            sources_queried.append("memory")
        if params.include_graph:
            sources_queried.append("graph")

        return CuratedContextResponse(
            query=params.query,
            sections=sections,
            metadata={
                "tokens_used": tokens_used,
                "tokens_budget": params.max_tokens,
                "sources_queried": sources_queried,
                "total_candidates": len(code_results["results"]) + len(memory_results["results"]) + len(graph_results["results"]),
                "results_returned": len(allocated),
                "execution_time_ms": elapsed_ms,
            },
        )

    def _search_code(self, analysis: Any) -> dict[str, Any]:
        try:
            fts_results = self._query_layer.search_code(analysis.fts_query, 30)

            # Also try direct symbol resolution
            symbol_results: list[dict[str, Any]] = []
            for candidate in analysis.symbol_candidates[:5]:
                resolved = self._resolver.resolve(candidate)
                for sym in resolved[:3]:
                    symbol_results.append({
                        "id": sym.id, "name": sym.name, "kind": sym.kind,
                        "file": sym.file_path, "line": sym.line, "signature": None,
                    })

            # Combine FTS + symbol results (mini-RRF)
            fts_items = [
                {"name": r["name"], "kind": r["kind"], "file": r["file_path"],
                 "line": r["start_line"], "signature": r.get("signature")}
                for r in fts_results
            ]
            combined = self._mini_rrf(fts_items, symbol_results)
            return {"source": "code", "results": combined[:20]}
        except Exception:
            return {"source": "code", "results": []}

    def _search_memory(self, analysis: Any) -> dict[str, Any]:
        try:
            query = " ".join(analysis.keywords[:5])
            cur = self._conn.execute(
                """SELECT id, content, summary, type, tags, created_at
                   FROM knowledge_entries
                   WHERE id IN (
                       SELECT rowid FROM knowledge_fts WHERE knowledge_fts MATCH ?
                   )
                   ORDER BY created_at DESC LIMIT 10""",
                (query,),
            )
            rows = cur.fetchall()
            results = []
            for row in rows:
                results.append({
                    "id": row[0],
                    "name": row[2] or (row[1][:50] if row[1] else "entry"),
                    "kind": row[3] or "memory",
                    "content": row[2] or (row[1][:200] if row[1] else ""),
                    "file": None, "line": None,
                })
            return {"source": "memory", "results": results}
        except Exception:
            return {"source": "memory", "results": []}

    def _expand_graph(self, top_symbols: list[dict[str, Any]]) -> dict[str, Any]:
        expanded: list[dict[str, Any]] = []
        seen: set[str] = set()

        for symbol in top_symbols:
            try:
                start_node = self._traverser.resolve_node(symbol.get("name", ""))
                if not start_node:
                    continue

                from ..graph.models import TraverseConfig
                results = self._traverser.traverse(start_node, TraverseConfig(
                    edge_types=["calls", "imports", "inherits"],
                    direction="both", max_depth=1, max_results=5,
                ))

                for r in results:
                    key = f"{r.node.name}:{r.node.file_path}"
                    if key in seen:
                        continue
                    seen.add(key)
                    expanded.append({
                        "id": r.node.id, "name": r.node.name,
                        "kind": r.node.kind, "file": r.node.file_path,
                        "line": r.node.start_line,
                        "relationship": f"{r.edge_type} {symbol.get('name', '')}",
                    })
            except Exception:
                continue

        return {"source": "graph", "results": expanded}

    def _mini_rrf(self, list_a: list[dict], list_b: list[dict]) -> list[dict]:
        k = 60
        scores: dict[str, dict[str, Any]] = {}

        for i, item in enumerate(list_a):
            key = f"{item.get('name', '')}:{item.get('file', '')}"
            scores[key] = {"score": 1 / (k + i), "item": item}

        for i, item in enumerate(list_b):
            key = f"{item.get('name', '')}:{item.get('file', '')}"
            if key in scores:
                scores[key]["score"] += 1 / (k + i)
            else:
                scores[key] = {"score": 1 / (k + i), "item": item}

        return [e["item"] for e in sorted(scores.values(), key=lambda x: x["score"], reverse=True)]

    def _format_sections(self, allocated: list[AllocatedResult]) -> list[ContextSection]:
        by_source: dict[str, list[ContextItem]] = {}

        for item in allocated:
            source = item.sources[0] if item.sources else "code"
            if source not in by_source:
                by_source[source] = []
            by_source[source].append(ContextItem(
                name=item.name, kind=item.kind, file=item.file,
                line=item.line, relevance=item.relevance_score,
                detail=item.detail, content=item.content,
                relationship=item.relationship,
            ))

        title_map = {"code": "Code Symbols", "memory": "Knowledge Base", "graph": "Related (Graph)"}
        return [
            ContextSection(title=title_map.get(source, source), source=source, items=items)
            for source, items in by_source.items()
        ]
